### DISCUSSION TITLE
Change from crypto++ to botan

### DISCUSSION THREAD

**Discussion by RobertSobral** [randombit/botan#4371] [Q&A] [answered]
I want to migrate my crypto++ code to Botan

Currently crypto++ is very outdated in relation to new methodologies, botan seems very up to date and has daily updates

I would like some information and tips on the best way I can convert my code

**Does Botan support Zlib like crypto++?**

for example I am working with a byte array and in crypto++ I use

```c
CryptoPP::ZlibDecompressor dec;
CryptoPP::ArraySource ar(src, sLen, true, new CryptoPP::Redirector(dec));
decompressor.Get(psrc, static_cast<size_t>(maxr));
```

I'm also encrypting a byte array, as they are bytes of files compressed in zlib, as far as I've tested, I can't increase the size of the bytes of files compressed in zlib, if I'm wrong, please let me know

What way does botan use to encrypt without using **AEAD_Mode** just normal AES? 

I have following code to decrypt:

```c
std::vector<uint8_t> decData;

CryptoPP::CTR_Mode<CryptoPP::AES>::Decryption dec;
dec.SetKeyWithIV(MyKey.data(), MyKey.size(), MyIV.data());
decData.resize(srcLen);

CryptoPP::ArraySource arraySrc(src, srcLen, true,
	new CryptoPP::StreamTransformationFilter(dec,
	new CryptoPP::ArraySink(decData.data(), decData.size())
	)
);


return decData;
```

**In the botan to clear the memory, `Botan::secure_scrub_memory` is used?**

**Comment by randombit** [comment, marked as answer]
> Does Botan support Zlib like crypto++?

Yes, but unlike Crypto++ zlib is based on the 3rd party `zlib` library rather than being implemented directly. So you must enable support at build time (`--with-zlib`), almost all distros/packagers use this option. You can see an example of usage in the [command line utility](https://github.com/randombit/botan/blob/master/src/cli/compress.cpp)

> What way does botan use to encrypt without using AEAD_Mode just normal AES?

So for AES in counter mode your snippet would look something like this (untested)

```
auto ctr = Botan::StreamCipher::create_or_throw("AES-128/CTR-BE");
ctr->set_key(MyKey.data(), MyKey.size());
ctr->set_iv(MyIv.data(), MyIv.size());

std::vector<uint8_t> decData;
decData.resize(srcLen);
ctr->cipher(src, decData.data(), srcLen);
return decData;

// or
std::vector<uint8_t> buf(src, src + srcLen);
ctr->cipher(buf);
return buf;
```

> In the botan to clear the memory, Botan::secure_scrub_memory is used?

Correct

**Comment by RobertSobral** [comment]
@randombit Thanks for support

Is there an AES-256/CTR mode?

Where can I find a list of existing options?

**Reply by randombit** [reply]
It is entirely generic so if there is a block cipher `X` then `X/CTR` works. Relevant doc regarding block ciphers is https://botan.randombit.net/handbook/api_ref/block_cipher.html#available-ciphers and CTR is https://botan.randombit.net/handbook/api_ref/stream_ciphers.html#ctr-be

**Comment by RobertSobral** [comment]
Code converted, thanks for the infos
